code/client/cl_http_windows.c

```diff
@@ -0,0 +1,162 @@
+static void DropIf(qboolean condition, const char *fmt, ...)
+{
+    char buffer[1024];
+
+    if (!condition)
+        return;
+
+    va_list argptr;
+    va_start(argptr, fmt);
+    Q_vsnprintf(buffer, sizeof(buffer), fmt, argptr);
+    va_end(argptr);
+
+    Com_Error(ERR_DROP, "Download Error: %s URL: %s", buffer, clc.downloadURL);
+}
+
+/*
+=================
+CL_HTTP_Init
+=================
+*/
+qboolean CL_HTTP_Init()
+{
+    OSVERSIONINFO osvi = {sizeof(OSVERSIONINFO)};
+    const char *windowsVersion = GetVersionEx(&osvi) ? va("Windows %lu.%lu (build %lu)",
+                                                          osvi.dwMajorVersion,
+                                                          osvi.dwMinorVersion,
+                                                          osvi.dwBuildNumber)
+                                                     : "Windows";
+
+    hInternet = InternetOpenA(
+        va("%s %s", Q3_VERSION, windowsVersion),
+        INTERNET_OPEN_TYPE_PRECONFIG, NULL, NULL, 0);
+
+    return hInternet != NULL;
+}
+
+/*
+=================
+CL_HTTP_Available
+=================
+*/
+qboolean CL_HTTP_Available()
+{
+    return hInternet != NULL;
+}
+
+/*
+=================
+CL_HTTP_Shutdown
+=================
+*/
+void CL_HTTP_Shutdown(void)
+{
+    if (hInternet)
+    {
+        InternetCloseHandle(hInternet);
+        hInternet = NULL;
+    }
+}
+
+/*
+=================
+CL_HTTP_BeginDownload
+=================
+*/
+void CL_HTTP_BeginDownload(const char *remoteURL)
+{
+    DWORD httpCode = 0;
+    DWORD contentLength = 0;
+    DWORD len = sizeof(httpCode);
+    DWORD zero = 0;
+    BOOL success;
+
+    hUrl = InternetOpenUrlA(hInternet, remoteURL, NULL, 0,
+                            INTERNET_FLAG_HYPERLINK |
+                                INTERNET_FLAG_NO_CACHE_WRITE |
+                                INTERNET_FLAG_NO_COOKIES |
+                                INTERNET_FLAG_NO_UI |
+                                INTERNET_FLAG_RESYNCHRONIZE |
+                                INTERNET_FLAG_RELOAD |
+                                INTERNET_FLAG_SECURE,
+                            0);
+
+    DropIf(hUrl == NULL, "InternetOpenUrlA failed %lu", GetLastError());
+
+    success = HttpQueryInfo(hUrl, HTTP_QUERY_STATUS_CODE | HTTP_QUERY_FLAG_NUMBER, &httpCode, &len, &zero);
+    DropIf(!success, "Get HTTP_QUERY_STATUS_CODE failed %lu", GetLastError());
+
+    DropIf(httpCode >= 400, "HTTP code %lu", httpCode);
+    DropIf(httpCode != 200, "Unhandled HTTP code %lu", httpCode);
+
+    success = HttpQueryInfo(hUrl, HTTP_QUERY_CONTENT_LENGTH | HTTP_QUERY_FLAG_NUMBER, &contentLength, &len, &zero);
+    DropIf(!success, "Get HTTP_QUERY_CONTENT_LENGTH failed %lu", GetLastError());
+
+    clc.downloadSize = (int)contentLength;
+    Cvar_SetValue("cl_downloadSize", clc.downloadSize);
+}
+
+/*
+=================
+CL_HTTP_PerformDownload
+=================
+*/
+qboolean CL_HTTP_PerformDownload(void)
+{
+    BYTE readBuffer[256 * 1024];
```

Comment:
Internet search says Windows programs default to a 1MB stack size. Pushing 256kb to the stack seems like a lot. Maybe use a smaller buffer like 16kb and call InternetReadFile() multiple times while bytesRead is more than 0? (I'm not particularly knowledgeable about this.)